### PR TITLE
Arch support - aka pacstrap, pacman and tests

### PR DESCRIPTION
**Pull request by evelikov**
This PR adds support for creating pacman based images (be that Arch or alike). Similar to the Debian variant we have two important actions - pacstrap and pacman.
 - pacstrap, requires pacman.conf and mirrorlist files
 - pacman is a simple list of packages, just like apt
 
 This PR supersedes https://github.com/go-debos/debos/pull/276 and https://github.com/go-debos/debos/pull/277, where the most significant changes are within the pacstrap action. In particular, we no longer write the config/mirrorlist files.

Considering that pacman is changing faster than debos, it makes sense to have the files as drop-ins instead of trying to hard-code them within debos itself.

Note: the final commit adds a simple CI file, which is not wired. Tips on how to achieve that would be appreciated.

### CODE REVIEW

**Comment by evelikov** [comment]
@obbardc feel free to close the superseded PRs - I don't have access.

File: actions/pacman_action.go
Comment (obbardc) [inline actions/pacman_action.go:30]: i think it'd be great to check in Verify if pacman is installed

But this can be a follow-up as I we should do the same check with all actions. Can you open an issue about that ?

File: actions/pacman_action.go
Comment (obbardc) [inline actions/pacman_action.go:34]:
```suggestion
	if err := c.Run("pacman", pacmanOptions...); err != nil {
```
to follow other actions

File: actions/pacstrap_action.go
Comment (obbardc) [inline actions/pacstrap_action.go:38]:
```suggestion
		return nil, fmt.Errorf("No config file set")
```

File: actions/pacstrap_action.go
Comment (obbardc) [inline actions/pacstrap_action.go:44]:
```suggestion
		return nil, fmt.Errorf("No mirror file set")
```

File: actions/pacstrap_action.go
Comment (obbardc) [inline actions/pacstrap_action.go:106]: what happens if this is run with `--disable-fakemachine` ? does it stamp on the host file ?

File: tests/arch/test.yaml
Comment (obbardc) [inline tests/arch/test.yaml:1]: to hook this up; you can:

1) modify `docker/Dockerfile` to install `pacman-package-manager` and `arch-install-scripts` (better put it in its own commented section a.la `# Install Arch dependencies`)
2) modify [CI definition](https://github.com/go-debos/debos/blob/main/.github/workflows/ci.yaml#L122) to add an arch test e.g:
```
          - { name: "arch", case: "arch" }
```

File: tests/arch/test.yaml
Comment (obbardc) [inline tests/arch/test.yaml:3]:
```suggestion
architecture: amd64
```
just hardcode it as we only support one arch anyway ?

File: actions/pacstrap_action.go
Comment (obbardc) [inline actions/pacstrap_action.go:52]: Should we check `context.Architecture==amd64` too ?

File: actions/pacman_action.go
Comment (obbardc) [inline actions/pacman_action.go:22]: Should we add `Verify` function to check `context.Architecture==amd64` ?

File: actions/pacman_action.go
Comment (evelikov) [reply actions/pacman_action.go:30]: Greed - a handful of actions make this assumption. Opened https://github.com/go-debos/debos/issues/379

File: tests/arch/test.yaml
Comment (evelikov) [reply tests/arch/test.yaml:3]: For all the arch specific questions: as it's written there is nothing preventing you from trying other architectures - be that arm, aarch or risc-v. Personally I haven't tested them and instead opted for being consistent with the Debian/debootstrap/apt codebase.

Can change that, but it seems very strange to single-out the Arch/pacstrap/pacman actions.

Comment (evelikov) [reply tests/arch/test.yaml:3]: If we are to hard-code things, should we do the same for Debian/debootstrap/apt?

File: actions/pacstrap_action.go
Comment (evelikov) [reply actions/pacstrap_action.go:106]: No idea to be honest - `fakemachine` is the only option that has been tried.

Unless this is a blocker, I'm leaning that we can error out if run w/o `fakemachine`. Are you comfortable with that?

Comment (obbardc) [reply actions/pacstrap_action.go:106]: > Unless this is a blocker, I'm leaning that we can error out if run w/o `fakemachine`. Are you comfortable with that?

You can do that as long as the CI tests pass ;-). We can add non-fakemachine support as a follow-up.

File: tests/arch/test.yaml
Comment (evelikov) [reply tests/arch/test.yaml:3]: https://archlinux32.org/ started since 32bit was dropped from the official distro - supports i486, i686 and pentium4
https://archlinuxarm.org/ has been a thing for 10+ years - supports armv7 and armv8 ... in the past it had armv6 support IIRC

Comment (evelikov) [reply tests/arch/test.yaml:3]: There is also WIP risc-v somewhere (sorry cannot find the link).

All of the above should just work with the PR, of course assuming debos/qemu runs on that platform.

File: actions/pacstrap_action.go
Comment (evelikov) [reply actions/pacstrap_action.go:106]: At a glance the least invasive approach is to:
 - save the context type within `debos.DebosContext`
 - in the `Verify` hooks for `pacstrap/pacman` will print "Unsupported backend"
 - where the `Run` hooks will silently error out

It does feel a bit hacky, but the alternative is quite CI churn heavy. The current testing has no "known failing" concept. 

Does the above sound reasonable or do you have a particular approach in mind?

File: tests/arch/test.yaml
Comment (evelikov) [reply tests/arch/test.yaml:1]: This doesn't quite work - `pacman-package-manager` is available only in `bookworm` and `sid`. The other package is present across the board.

Comment (obbardc) [reply tests/arch/test.yaml:1]: bump the container baseline to bookworm for testing this PR, we can cherry pick that into another PR later on.

File: actions/pacstrap_action.go
Comment (obbardc) [reply actions/pacstrap_action.go:106]: I think it'd be great to first answer the question of what could happen to the user's machine if pacman/pacstrap action is run with no fakemachine

but your approach sounds ok; but instead of having to add a separate field to context and running in `Verify` you could just error out in `PreNoMachine` instead ;-)

File: tests/arch/test.yaml
Comment (obbardc) [reply tests/arch/test.yaml:3]: right - thanks for the history. I thought arch was arm64 only.

So, the named architectures here should just build ootb we can just resolve this.

File: docker/Dockerfile
Comment (obbardc) [inline docker/Dockerfile:91]: Just need to change the container baseline from `bullseye` to `bookworm` - since `pacman-package-manager` is not available in bullseye.

If it works, we can ask the author to backport the package to bullseye-backports and install from there, or just stick with bumping the container baseline. If you can do that in a commit on top, we can always decide that in another PR.

File: actions/pacstrap_action.go
Comment (evelikov) [reply actions/pacstrap_action.go:110]: Yes, when run without fakemachine the current pacstrap stage will try and override the system files - pacman.conf and mirrorlist. Assuming debos has enough permissions (run as root or otherwise) it will succeed.

Added an error out in `PreNoMachine` although that is causing the existing CI to choke. Any pointers would be apprecaited.

File: docker/Dockerfile
Comment (evelikov) [reply docker/Dockerfile:91]: Bumped - that does cause debian and aperitis tests to fail. Let me know what you'd prefer - omit the bump, fix those tests, other. 

Once the Arch/pacstrap/pacman CI finishes that it - it'll take some time.

File: actions/pacstrap_action.go
Comment (evelikov) [reply actions/pacstrap_action.go:52]: As per the CPU architecture discussion above, the check should not exist.

File: .github/workflows/ci.yaml
Comment (obbardc) [inline .github/workflows/ci.yaml:124]: if you just want arch tests to run in a fakemachine and not in the `--disable-fakemachine` case, you can do this instead:

```
include:
  - backend: { name: "arch", backend: "--fakemachine-backend=uml" }
    test: { name: "arch", case: "arch" }
```

File: docker/Dockerfile
Comment (evelikov) [reply docker/Dockerfile:91]: Seems like the bump to bookworm is breaking DNS/name resolution for all distros - Debian, Aperitis and Arch.

Comment (obbardc) [reply docker/Dockerfile:91]: sorry I forgot to mention - in that update you will also need to install the "systemd-resolved" package

File: actions/pacstrap_action.go
Comment (evelikov) [reply actions/pacstrap_action.go:110]: Fixed with the "include" snippet that you're shared. Thank you :bowing_man:

File: docker/Dockerfile
Comment (evelikov) [reply docker/Dockerfile:92]: Fixed thanks.

**Comment by evelikov** [comment]
Split the bookworm uprev to https://github.com/go-debos/debos/pull/382

The failures are due to json formatting variance. At a glance the proper solution is to call `jq .` on both files prior to feeding them into `diff`. Will open a separate PR in a second

**Comment by obbardc** [comment]
@evelikov We're tracking pulling `pacman-package-manager` into `bullseye-backports` under https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1023386

**Comment by evelikov** [comment]
Glancing at the Debian bug - seems like the backport is not so straight-forward.

To unblock us, we can build pacman locally in the short term. Then switch to the official package as it becomes available. @obbardc what do you think?

**Comment by obbardc** [comment]
> Glancing at the Debian bug - seems like the backport is not so straight-forward.
> 
> To unblock us, we can build pacman locally in the short term. Then switch to the official package as it becomes available. @obbardc what do you think?

That sounds like a great plan!

**Comment by evelikov** [comment]
Was quite sick last week and my TODO stack is quite high, so I don't think I'll get to it this week.
The upstream discussion seems nearly resolved so, it might be better to revisit this next week/other after.

Will flip this to draft, so it doesn't get merged accidentally - the mirrorlist URLs need fixing.

File: actions/pacstrap_action.go
Comment (obbardc) [inline actions/pacstrap_action.go:114]: nit: pacman-key lowercase

File: actions/pacstrap_action.go
Comment (obbardc) [inline actions/pacstrap_action.go:120]: nit: pacman-key lowercase

File: actions/pacstrap_action.go
Comment (obbardc) [inline actions/pacstrap_action.go:126]: nit: pacstrap lowercase

**Review by obbardc — CHANGES_REQUESTED** [review]

File: actions/pacstrap_action.go
Comment (obbardc) [reply actions/pacstrap_action.go:126]: this is the last nit remaining @evelikov :p

Comment (evelikov) [reply actions/pacstrap_action.go]: Sorry about that. Think I've got them all covered now :crossed_fingers:

**Comment by evelikov** [comment]
Quick look at the CI shows something quite odd:

We enable `bullseye-backports` before we do anything Arch related.

Yet `arch-install-scripts v23-3` from `bullseye` gets installed which [is missing pacstrap](https://packages.debian.org/bullseye/all/arch-install-scripts/filelist). Instead I would expect the one from backports to be installed (v28-1~bpo11+1) which [has the required tool](https://packages.debian.org/bullseye-backports/all/arch-install-scripts/filelist).

@obbardc any ideas how can we convince `apt` to use the backports variants for the whole stack?

**Comment by obbardc** [comment]
> Quick look at the CI shows something quite odd:
> 
> We enable `bullseye-backports` before we do anything Arch related.
> 
> Yet `arch-install-scripts v23-3` from `bullseye` gets installed which [is missing pacstrap](https://packages.debian.org/bullseye/all/arch-install-scripts/filelist). Instead I would expect the one from backports to be installed (v28-1~bpo11+1) which [has the required tool](https://packages.debian.org/bullseye-backports/all/arch-install-scripts/filelist).
> 
> @obbardc any ideas how can we convince `apt` to use the backports variants for the whole stack?

Ah, that explains things :-)! Please try `apt-get install -t bullseye-backports <package_name>`

**Comment by evelikov** [comment]
The issue we're seeing here is that one of the keys have expired and thus validation fails.

The Debian `archlinux-keyring` package is tad old - upstream has a new release to fix that.

**Comment by obbardc** [comment]
> The issue we're seeing here is that one of the keys have expired and thus validation fails.
> 
> The Debian `archlinux-keyring` package is tad old - upstream has a new release to fix that.

I opened an upstream ticket (note that you could also do the same in future if other packages appear outdated)
It's tracked at: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1026080
For now to move this patch forward, since the backport may take weeks/months, can you somehow install the keys manually and put the bug link in the source code ?

**Comment by evelikov** [comment]
Sure thing - pulled the keyring installation locally. I'm wondering if it makes sense to actually keep it all together, since there can be quite notable delays until an updated package lands in backports.

If I understand correctly, there is a 1-2 week delay between the new version migrating from unstable to testing. Only after it lands in testing it's allowed in backports. During which time the CI will be broken and as result bors won't be able to merge.

**Comment by obbardc** [comment]
Nice work!

bors r+

**Comment by evelikov** [comment]
Now that is a much appreciated Xmas/NYE present. Thanks o/

**Pull request merged**
